### PR TITLE
Iy config manager

### PR DESCRIPTION
**Pull request by iiman**
## Approvals

- [x] Verified
- [ ] @hebbo
- [ ] @ptolemy428
- [x] @iiman

`configmanager`: a module for loading stack configuration according to a specified order.

- A  stack configuration consists of at least `template` and `account` settings. 

    --  `template` includes settings whose values remains the same irrespective of in which account the stack is created.  

   -- `account` contains account specific settings such as kms keys.

  -- Stack-specific keys, like `db_stack_name`, can be provided in a seperate file. Furthermore the settings of `template` and `account` can be overridden in the separate file.

- The loading order is provided via a file, default name `_order`

```
_template.yml
_account.yml
mystack.yml
```

`configmanager` loads the content of the file in the first line, then the content of the file in the second, and so forth until it reaches the last file. If a setting is specified by more than one file, the value of that the setting will be taken from the file that was read the last.

### CODE REVIEW

File: tests/unit/utils/test_configmanager.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import unittest
+import tempfile
+import os
+
+from cloudformationcore.utils.configmanager import generate_config
+
+
+class TestValidator(unittest.TestCase):
+    def test_should_generate_config(self):
+        tmpdir = tempfile.mkdtemp()
```
Comment (ptolemy428) [inline tests/unit/utils/test_configmanager.py]: It's perfectly ok to have testing configs stored in the testing director to support the unit testing logics, which makes the test code easier to read and less OS dependent. It's usually not a recommended way to dynamically generate test data and persist to files as part of the unit tests. which also needs to be cleaned up but it's not really a concern of the tests.

File: cloudformationcore/utils/configmanager.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import os
+
+from yamlhelper import load_config
+
+
+def generate_config(config_files_path, configs_loading_order):
```
Comment (ptolemy428) [inline cloudformationcore/utils/configmanager.py]: How do we run this in debug mode and how does the config merges get logged for people to see? Ideally it should be logged to stdout.

Comment (iiman) [inline cloudformationcore/utils/configmanager.py]: Suppose you want to see the config for `app` stack:

` bin/docker/cloudformation-core  python roles/cfn/files/app/app_main.py  -region usw2 -stack_prefix sample  -d -config_folder  roles/cfn/files/app/config -account_config account/dev-usw2.yml`

File: cloudformationcore/utils/configmanager.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        # skips files if they dont exist
+        try:
+            logger.debug('loading configuration from %s' % (i))
+            partial_config = load_config(os.path.join(config_folder, i))
+
+            common_keys = _get_overwritten_keys(complete_config,
+                                                partial_config)
+            if common_keys:
+                logger.warn('following keys will be overwritten: \n %s' %
+                            pprint.pformat(common_keys))
```
Comment (ptolemy428) [inline cloudformationcore/utils/configmanager.py]: can we print out the value as well.

File: cloudformationcore/utils/configmanager.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                              stack_config='',
+                              configs_ordering_file=''):
+    try:
+        validate_config(configs_ordering_file)
+        with open(configs_ordering_file, 'r') as f:
+            loading_order = f.readlines()
+    except ValueError:
+        loading_order = _get_default_order(account_config, stack_config)
+
+    return [x.strip('\n') for x in loading_order if x.strip()]
```
Comment (ptolemy428) [inline cloudformationcore/utils/configmanager.py]: config as code.

File: cloudformationcore/utils/validator.py [hunk 5]
```diff
@@ -6,7 +6,13 @@ def validate_args():
...
+    parser.add_argument(
+        '-config', required=False, type=str)  # backward compatability
+    parser.add_argument('-config_folder', required=True, type=str)
+    parser.add_argument('-config_order', required=False, type=str)
+    parser.add_argument('-account_config', required=False, type=str)
+    parser.add_argument('-stack_config', required=False, type=str)
+    parser.add_argument('-d', '--debug', action='store_true', required=False)
     args = parser.parse_args()
 
     return args
```
Comment (ptolemy428) [inline cloudformationcore/utils/validator.py]: an entrypoing would be great __main__

File: cloudformationcore/utils/configmanager.py (hunk 3 above)
Comment (iiman) [inline cloudformationcore/utils/configmanager.py]: As discussed, the values are already printed in the next line

**Comment by iiman** [comment]
[unit test of confimanager](https://trello.com/c/QUk02z5A/537-unit-test-of-confimanager)

File: tests/unit/utils/test_configmanager.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            stack_config='test-prefix-account-region.yml')['instance_type']
+        self.assertEqual(expected, actual)
+
+        expected = 't2.medium'
+        actual = generate_config(
+            'tests/unit/utils/config',
+            account_config='account/account-region.yml',
+            stack_config='test-overwrite-prefix-account-region.yml')[
+                'instance_type']
+        self.assertEqual(expected, actual)
```
Comment (hebbo) [inline tests/unit/utils/test_configmanager.py]: Maybe we can add an extra test to make sure values set in the account/region config file are also overwritten?

Comment (iiman) [inline tests/unit/utils/test_configmanager.py]: This is done now.

**Pull request merged**